File: binding/src/libqtoolbar.cpp
#include <QAction>
#include <QActionEvent>
#include <QEvent>
#include <QIcon>
#include <QPaintEvent>
#include <QPoint>
#include <QRect>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QStyleOptionToolBar>
#include <QToolBar>
#include <QWidget>
#include <qtoolbar.h>
#include "libqtoolbar.h"
#include "libqtoolbar.hxx"

QToolBar* QToolBar_new(QWidget* parent) {
	 return new VirtualQToolBar(parent);
}

QToolBar* QToolBar_new2(const libqt_string title) {
	 return new VirtualQToolBar(QString::fromUtf8(title.data, title.len));
}

QToolBar* QToolBar_new3() {
	 return new VirtualQToolBar();
}

QToolBar* QToolBar_new4(const libqt_string title, QWidget* parent) {
	 return new VirtualQToolBar(QString::fromUtf8(title.data, title.len), parent);
}

libqt_string QToolBar_Tr(const char* s) {
	QString _ret = QToolBar::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QToolBar_SetMovable(QToolBar* self, bool movable) {
	self->setMovable(movable);
}

bool QToolBar_IsMovable(const QToolBar* self) {
	return self->isMovable();
}

void QToolBar_SetAllowedAreas(QToolBar* self, int areas) {
	self->setAllowedAreas(static_cast<QFlags<Qt::ToolBarArea>>(areas));
}

int QToolBar_AllowedAreas(const QToolBar* self) {
	return self->allowedAreas();
}

bool QToolBar_IsAreaAllowed(const QToolBar* self, int area) {
	return self->isAreaAllowed(static_cast<Qt::ToolBarArea>(area));
}

void QToolBar_SetOrientation(QToolBar* self, int orientation) {
	self->setOrientation(static_cast<Qt::Orientation>(orientation));
}

int QToolBar_Orientation(const QToolBar* self) {
	return self->orientation();
}

void QToolBar_Clear(QToolBar* self) {
	self->clear();
}

QAction* QToolBar_AddSeparator(QToolBar* self) {
	return self->addSeparator();
}

QAction* QToolBar_InsertSeparator(QToolBar* self, QAction* before) {
	return self->insertSeparator(before);
}

QAction* QToolBar_AddWidget(QToolBar* self, QWidget* widget) {
	return self->addWidget(widget);
}

QAction* QToolBar_InsertWidget(QToolBar* self, QAction* before, QWidget* widget) {
	return self->insertWidget(before, widget);
}

QRect* QToolBar_ActionGeometry(const QToolBar* self, QAction* action) {
	return new QRect(self->actionGeometry(action));
}

QAction* QToolBar_ActionAt(const QToolBar* self, const QPoint* p) {
	return self->actionAt(*p);
}

QAction* QToolBar_ActionAt2(const QToolBar* self, int x, int y) {
	return self->actionAt(x, y);
}

QAction* QToolBar_ToggleViewAction(const QToolBar* self) {
	return self->toggleViewAction();
}

QSize* QToolBar_IconSize(const QToolBar* self) {
	return new QSize(self->iconSize());
}

int QToolBar_ToolButtonStyle(const QToolBar* self) {
	return self->toolButtonStyle();
}

QWidget* QToolBar_WidgetForAction(const QToolBar* self, QAction* action) {
	return self->widgetForAction(action);
}

bool QToolBar_IsFloatable(const QToolBar* self) {
	return self->isFloatable();
}

void QToolBar_SetFloatable(QToolBar* self, bool floatable) {
	self->setFloatable(floatable);
}

bool QToolBar_IsFloating(const QToolBar* self) {
	return self->isFloating();
}

void QToolBar_SetIconSize(QToolBar* self, const QSize* iconSize) {
	self->setIconSize(*iconSize);
}

void QToolBar_SetToolButtonStyle(QToolBar* self, int toolButtonStyle) {
	self->setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(toolButtonStyle));
}

void QToolBar_ActionTriggered(QToolBar* self, QAction* action) {
	self->actionTriggered(action);
}

void QToolBar_Connect_ActionTriggered(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, QAction*) = reinterpret_cast<void (*)(QToolBar*, QAction*)>(slot);
    QToolBar::connect(self, &QToolBar::actionTriggered, [self, slotFunc](QAction* action) {
	slotFunc(self, action);
    });
}

void QToolBar_MovableChanged(QToolBar* self, bool movable) {
	self->movableChanged(movable);
}

void QToolBar_Connect_MovableChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, bool) = reinterpret_cast<void (*)(QToolBar*, bool)>(slot);
    QToolBar::connect(self, &QToolBar::movableChanged, [self, slotFunc](bool movable) {
	slotFunc(self, movable);
    });
}

void QToolBar_AllowedAreasChanged(QToolBar* self, int allowedAreas) {
	self->allowedAreasChanged(static_cast<QFlags<Qt::ToolBarArea>>(allowedAreas));
}

void QToolBar_Connect_AllowedAreasChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, int) = reinterpret_cast<void (*)(QToolBar*, int)>(slot);
    QToolBar::connect(self, &QToolBar::allowedAreasChanged, [self, slotFunc](Qt::ToolBarAreas allowedAreas) {
	slotFunc(self, allowedAreas);
    });
}

void QToolBar_OrientationChanged(QToolBar* self, int orientation) {
	self->orientationChanged(static_cast<Qt::Orientation>(orientation));
}

void QToolBar_Connect_OrientationChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, int) = reinterpret_cast<void (*)(QToolBar*, int)>(slot);
    QToolBar::connect(self, &QToolBar::orientationChanged, [self, slotFunc](Qt::Orientation orientation) {
	slotFunc(self, orientation);
    });
}

void QToolBar_IconSizeChanged(QToolBar* self, const QSize* iconSize) {
	self->iconSizeChanged(*iconSize);
}

void QToolBar_Connect_IconSizeChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, const QSize*) = reinterpret_cast<void (*)(QToolBar*, const QSize*)>(slot);
    QToolBar::connect(self, &QToolBar::iconSizeChanged, [self, slotFunc](const QSize& iconSize) {
	slotFunc(self, iconSize);
    });
}

void QToolBar_ToolButtonStyleChanged(QToolBar* self, int toolButtonStyle) {
	self->toolButtonStyleChanged(static_cast<Qt::ToolButtonStyle>(toolButtonStyle));
}

void QToolBar_Connect_ToolButtonStyleChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, int) = reinterpret_cast<void (*)(QToolBar*, int)>(slot);
    QToolBar::connect(self, &QToolBar::toolButtonStyleChanged, [self, slotFunc](Qt::ToolButtonStyle toolButtonStyle) {
	slotFunc(self, toolButtonStyle);
    });
}

void QToolBar_TopLevelChanged(QToolBar* self, bool topLevel) {
	self->topLevelChanged(topLevel);
}

void QToolBar_Connect_TopLevelChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, bool) = reinterpret_cast<void (*)(QToolBar*, bool)>(slot);
    QToolBar::connect(self, &QToolBar::topLevelChanged, [self, slotFunc](bool topLevel) {
	slotFunc(self, topLevel);
    });
}

void QToolBar_VisibilityChanged(QToolBar* self, bool visible) {
	self->visibilityChanged(visible);
}

void QToolBar_Connect_VisibilityChanged(QToolBar* self, intptr_t slot) {
    void (*slotFunc)(QToolBar*, bool) = reinterpret_cast<void (*)(QToolBar*, bool)>(slot);
    QToolBar::connect(self, &QToolBar::visibilityChanged, [self, slotFunc](bool visible) {
	slotFunc(self, visible);
    });
}

libqt_string QToolBar_Tr2(const char* s, const char* c) {
	QString _ret = QToolBar::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QToolBar_Tr3(const char* s, const char* c, int n) {
	QString _ret = QToolBar::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QToolBar_Delete(QToolBar* self) {
    delete self;
}

File: binding/include/libqtoolbar.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qtoolbar.html
pub const qtoolbar = struct {

    /// New constructs a new QToolBar object.
    pub fn New(parent: ?*anyopaque) QtC.QToolBar {
        return qtc.QToolBar_new(@ptrCast(parent));
    }


    /// New2 constructs a new QToolBar object.
    pub fn New2(title: []const u8) QtC.QToolBar {
        const title_str = qtc.libqt_string{
    .len = title.len,
    .data = title.ptr,
};
return qtc.QToolBar_new2(title_str);
    }


    /// New3 constructs a new QToolBar object.
    pub fn New3() QtC.QToolBar {
        return qtc.QToolBar_new3();
    }


    /// New4 constructs a new QToolBar object.
    pub fn New4(title: []const u8, parent: ?*anyopaque) QtC.QToolBar {
        const title_str = qtc.libqt_string{
    .len = title.len,
    .data = title.ptr,
};
return qtc.QToolBar_new4(title_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBar_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbar.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetMovable(self: ?*anyopaque, movable: bool) void {
        qtc.QToolBar_SetMovable(@ptrCast(self), movable);
    }

    pub fn IsMovable(self: ?*anyopaque, ) bool {
        return qtc.QToolBar_IsMovable(@ptrCast(self));
    }

    pub fn SetAllowedAreas(self: ?*anyopaque, areas: i32) void {
        qtc.QToolBar_SetAllowedAreas(@ptrCast(self), @intCast(areas));
    }

    pub fn AllowedAreas(self: ?*anyopaque, ) i32 {
        return qtc.QToolBar_AllowedAreas(@ptrCast(self));
    }

    pub fn IsAreaAllowed(self: ?*anyopaque, area: i32) bool {
        return qtc.QToolBar_IsAreaAllowed(@ptrCast(self), @intCast(area));
    }

    pub fn SetOrientation(self: ?*anyopaque, orientation: i32) void {
        qtc.QToolBar_SetOrientation(@ptrCast(self), @intCast(orientation));
    }

    pub fn Orientation(self: ?*anyopaque, ) i32 {
        return qtc.QToolBar_Orientation(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QToolBar_Clear(@ptrCast(self));
    }

    pub fn AddSeparator(self: ?*anyopaque, ) QtC.QAction {
        return qtc.QToolBar_AddSeparator(@ptrCast(self));
    }

    pub fn InsertSeparator(self: ?*anyopaque, before: ?*anyopaque) QtC.QAction {
        return qtc.QToolBar_InsertSeparator(@ptrCast(self), @ptrCast(before));
    }

    pub fn AddWidget(self: ?*anyopaque, widget: ?*anyopaque) QtC.QAction {
        return qtc.QToolBar_AddWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn InsertWidget(self: ?*anyopaque, before: ?*anyopaque, widget: ?*anyopaque) QtC.QAction {
        return qtc.QToolBar_InsertWidget(@ptrCast(self), @ptrCast(before), @ptrCast(widget));
    }

    pub fn ActionGeometry(self: ?*anyopaque, action: ?*anyopaque) QtC.QRect {
        return qtc.QToolBar_ActionGeometry(@ptrCast(self), @ptrCast(action));
    }

    pub fn ActionAt(self: ?*anyopaque, p: ?*anyopaque) QtC.QAction {
        return qtc.QToolBar_ActionAt(@ptrCast(self), @ptrCast(p));
    }

    pub fn ActionAt2(self: ?*anyopaque, x: i32, y: i32) QtC.QAction {
        return qtc.QToolBar_ActionAt2(@ptrCast(self), x, y);
    }

    pub fn ToggleViewAction(self: ?*anyopaque, ) QtC.QAction {
        return qtc.QToolBar_ToggleViewAction(@ptrCast(self));
    }

    pub fn IconSize(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QToolBar_IconSize(@ptrCast(self));
    }

    pub fn ToolButtonStyle(self: ?*anyopaque, ) i32 {
        return qtc.QToolBar_ToolButtonStyle(@ptrCast(self));
    }

    pub fn WidgetForAction(self: ?*anyopaque, action: ?*anyopaque) QtC.QWidget {
        return qtc.QToolBar_WidgetForAction(@ptrCast(self), @ptrCast(action));
    }

    pub fn IsFloatable(self: ?*anyopaque, ) bool {
        return qtc.QToolBar_IsFloatable(@ptrCast(self));
    }

    pub fn SetFloatable(self: ?*anyopaque, floatable: bool) void {
        qtc.QToolBar_SetFloatable(@ptrCast(self), floatable);
    }

    pub fn IsFloating(self: ?*anyopaque, ) bool {
        return qtc.QToolBar_IsFloating(@ptrCast(self));
    }

    pub fn SetIconSize(self: ?*anyopaque, iconSize: ?*anyopaque) void {
        qtc.QToolBar_SetIconSize(@ptrCast(self), @ptrCast(iconSize));
    }

    pub fn SetToolButtonStyle(self: ?*anyopaque, toolButtonStyle: i32) void {
        qtc.QToolBar_SetToolButtonStyle(@ptrCast(self), @intCast(toolButtonStyle));
    }

    pub fn ActionTriggered(self: ?*anyopaque, action: ?*anyopaque) void {
        qtc.QToolBar_ActionTriggered(@ptrCast(self), @ptrCast(action));
    }

    pub fn OnActionTriggered(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_Connect_ActionTriggered(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MovableChanged(self: ?*anyopaque, movable: bool) void {
        qtc.QToolBar_MovableChanged(@ptrCast(self), movable);
    }

    pub fn OnMovableChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QToolBar_Connect_MovableChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn AllowedAreasChanged(self: ?*anyopaque, allowedAreas: i32) void {
        qtc.QToolBar_AllowedAreasChanged(@ptrCast(self), @intCast(allowedAreas));
    }

    pub fn OnAllowedAreasChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBar_Connect_AllowedAreasChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn OrientationChanged(self: ?*anyopaque, orientation: i32) void {
        qtc.QToolBar_OrientationChanged(@ptrCast(self), @intCast(orientation));
    }

    pub fn OnOrientationChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBar_Connect_OrientationChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn IconSizeChanged(self: ?*anyopaque, iconSize: ?*anyopaque) void {
        qtc.QToolBar_IconSizeChanged(@ptrCast(self), @ptrCast(iconSize));
    }

    pub fn OnIconSizeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_Connect_IconSizeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ToolButtonStyleChanged(self: ?*anyopaque, toolButtonStyle: i32) void {
        qtc.QToolBar_ToolButtonStyleChanged(@ptrCast(self), @intCast(toolButtonStyle));
    }

    pub fn OnToolButtonStyleChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBar_Connect_ToolButtonStyleChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TopLevelChanged(self: ?*anyopaque, topLevel: bool) void {
        qtc.QToolBar_TopLevelChanged(@ptrCast(self), topLevel);
    }

    pub fn OnTopLevelChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QToolBar_Connect_TopLevelChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn VisibilityChanged(self: ?*anyopaque, visible: bool) void {
        qtc.QToolBar_VisibilityChanged(@ptrCast(self), visible);
    }

    pub fn OnVisibilityChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QToolBar_Connect_VisibilityChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ActionEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_ActionEvent(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnActionEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_OnActionEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseActionEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_QBaseActionEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn ChangeEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_ChangeEvent(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnChangeEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_OnChangeEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseChangeEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_QBaseChangeEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn PaintEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_PaintEvent(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_OnPaintEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEvent(self: ?*anyopaque, event: ?*anyopaque) void {
        qtc.QToolBar_QBasePaintEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn Event(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QToolBar_Event(@ptrCast(self), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QToolBar_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QToolBar_QBaseEvent(@ptrCast(self), @ptrCast(event));
    }

    pub fn InitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QToolBar_InitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    /// Allows for overriding the related default method
    pub fn OnInitStyleOption(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBar_OnInitStyleOption(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseInitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QToolBar_QBaseInitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBar_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbar.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBar_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbar.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QToolBar_Delete(@ptrCast(self));
    }
};
